validate.js

```diff
@@ -77,28 +78,32 @@ export default class Validate extends Validates {
   }
 
   /**
-   * React lifecycle handler called immediately before the component's initial render.
+   * React lifecycle handler called immediately after the component's initial render.
    */
-  componentWillMount() {
+  componentDidMount() {
     // Update the handlers with the initial state
     this.onValidChange(this.validates);
   }
 
   /**
-   * React lifecycle handler called when a component is receiving new props.
+   * React lifecycle handler called when a component is mounting or receiving new props.
    *
-   * @param {Object} nextProps Component's new props.
+   * @param {Object} props Component's new props.
+   * @param {Object} state Component's current state.
+   * @returns {Object} updated state object.
    */
-  componentWillReceiveProps(nextProps) {
-    const { validate } = nextProps;
-    if (validate === this.props.validate) {
+  static getDerivedStateFromProps(props, state) {
```

Comment:
I am pretty sure this will not work because `getDerivedStateFromProps` is called multiple times in the component's lifecycle, not simply when it receives props. 

[This post](https://blog.logrocket.com/the-new-react-lifecycle-methods-in-plain-approachable-language-61a2105859f3) outlines the new lifecycle well. [This other post](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) goes into details about best practices for `getDerivedStateFromProps`. 

Perhaps we can accomplish this without `getDerivedStateFromProps?

Reply:
I may be completely incorrect in my thinking here, but from what I understand from [this lifecycle diagram] (https://twitter.com/dan_abramov/status/981712092611989509?lang=en) from Dan Abramov, it appears that `getDerivedStateFromProps` is called on initial mount and when props have changed. Based on that knowledge, the reason I went with this approach was so that we could check if the prop value had changed and if not, not update state nor re-render. 

Alternative approaches considered were:
- Check in `shouldComponentUpdate`, which I believe would be called more often than `getDerivedStateFromProps` since it happens on both `state` _and_ `prop` changes.
- Use `componentDidUpdate` in conjunction with the `constructor` to check, but (if I understand correctly) this happens post-render which, in the event of a change, would cause a re-render.

All that said, perhaps none of this is needed. I simply attempted a migration, but do we anticipate the `validate` function prop to change after the initial mount? If not, we can dispense with this entire block of code. Thoughts?

Reply:
We must go deeper. 

The post you linked to is the lifecycle diagram circa `react@16.3` – but `react@16.4` introduced the lifecycle diagram you can find [here](https://medium.com/@_ChrisBradshaw/blogdidmount-2018-guide-to-react-component-lifecycle-methods-1614e0bbe80a).

![react-164-lifecycle](https://user-images.githubusercontent.com/4624/53843017-22d7d280-3f6f-11e9-866b-b24a34ff9704.png)

As you can see the difference is that you'll get calls to `getDerivedStateFromProps` on calls to `setState` – this extra call is where I believe this logic will break down.

Reply:
Ah, I see the issue now. I'll think on that a bit more and come up with a different approach.

Reply:
Also, I need to correct my previous comment:

- `componentDidUpdate` could potentially work here if we utilize the same `setState` call:
> React ensures that any setState calls that happen during componentDidMount and componentDidUpdate are flushed before the user sees the updated UI.

Given that additional information, I'll try that path.

Reply:
@dustinsavery -- did you read [the second article](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) that @indexzero mentioned? At the bottom, it covers alternatives to `getDerivedStateFromProps` for memoization cases like the one here.

Reply:
I've run a number of tests for each of the options and here is what I've found:
- `componentDidUpdate`: By checking `prevProps.validate !== this.props.validate` and executing the function to update the validity, I was able to accomplish the expected behavior. When the `validate` function is swapped, it calls `componentDidUpdate`, re-evaluates the validity, and updates state. And since this is behind the above if check, it doesn't execute unless the function has changed. However, to accomodate this behavior (in conjuction with [this current logic](https://github.com/godaddy/react-validation-context/blob/master/validate.js#L111-L115)), I extracted out the current logic into a new function to keep it DRY. This seems to better than the current approach as it accomplishes the same thing, but without tracking the function in state.
- `getDerivedStateFromProps` (current implementation in PR): This was called on every update cycle (as expected per the diagram). However, since we do an early abort if the method hasn't changed, it didn't cause an increase in render cycles. The downside with this approach is we have to track the function in state, which is not ideal.
- `memoize`/`componentDidUpdate`: I saw no noticeable difference in performance or number of calls with this method from the plain `componentDidUpdate` option.

Given these findings, unless anyone objects, I think going with option 1 makes the most sense.